lang/cs/Org.Apache.REEF.Common/Telemetry/MetricsService.cs

```diff
@@ -65,13 +89,64 @@ public void OnNext(IContextMessage contextMessage)
                     //// For the counters from multiple evaluators with the same counter name, the value should be aggregated here
                     //// We also need to consider failure cases.  
                     _counters[counter.Name] = counter;
+
+                    if (_totalIncrementSinceLastSink.ContainsKey(counter.Name))
+                    {
+                        _totalIncrementSinceLastSink[counter.Name] += counter.Value - c.Value;
+                    }
+                    else
+                    {
+                        _totalIncrementSinceLastSink.Add(counter.Name, counter.Value - c.Value);
+                    }
                 }
                 else
                 {
                     _counters.Add(counter.Name, counter);
+                    _totalIncrementSinceLastSink.Add(counter.Name, counter.Value);
                 }
 
-                Logger.Log(Level.Verbose, "Counter name: {0}, value: {1}, description: {2}, time: {3}.", counter.Name, counter.Value, counter.Description, new DateTime(counter.Timestamp));
+                Logger.Log(Level.Verbose, "Counter name: {0}, value: {1}, description: {2}, time: {3},  incrementSinceLastSink: {4}.", 
+                    counter.Name, counter.Value, counter.Description, new DateTime(counter.Timestamp), _totalIncrementSinceLastSink[counter.Name]);
+            }
+
+            if (TriggerSink())
+            {
+                SinkCounters();
+                _totalIncrementSinceLastSink.Clear();
+            }
+        }
+
+        /// <summary>
+        /// The condition that triggers the sink. The condition can be modified later.
+        /// </summary>
+        /// <returns></returns>
+        private bool TriggerSink()
+        {
+            return _totalIncrementSinceLastSink.Sum(e => e.Value) > _counterSinkThreshold;
+        }
+
+        /// <summary>
+        /// Preparing data and call Sink 
+        /// </summary>
+        private void SinkCounters()
+        {
+            var set = new HashSet<KeyValuePair<string, string>>();
+            foreach (var c in _counters)
+            {
+                set.Add(new KeyValuePair<string, string>(c.Key, c.Value.Value.ToString()));
+            }
+
+            Sink(set);
+        }
+
+        /// <summary>
+        /// Call each Sink to sink the data in the counters
+        /// </summary>
+        private void Sink(ISet<KeyValuePair<string, string>> set)
+        {
+            foreach (var s in _metricsSinks)
+            {
+                s.Sink(set);
             }
```

Comment:
Good point. I will add try/catch. regarding timeout, we have no idea how long it would take. In side each sink, if it has any IO or network timeout, it should throw exception.